src/Validation/Factory.php

```diff
@@ -0,0 +1,76 @@
+<?php
+/* Copyright (c) 2017 Richard Klees <richard.klees@concepts-and-training.de> Extended GPL, see docs/LICENSE */
+
+namespace ILIAS\Validation;
+
+/**
+ * Factory for creating restrictions.
+ */
+class Factory {
+	// COMBINATORS
+
+	/**
+	 * Get a restriction that sequentially checks the supplied restrictions.
+	 *
+	 * The new restriction tells the problem of the first violated restriction.
+	 *
+	 * @param   Restriction[]   $others
+	 * @return  Restriction
+	 */
+	public function seq(array $others);
+
+	/**
+	 * Get a restriction that checks the supplied restrictions in parallel.
+	 *
+	 * The new restriction tells the problems of all violated restrictions.
+	 *
+	 * @param   Restriction[]   $others
+	 * @return	Restriction
+	 */
+	public function par(array $others);
+
+	/**
+	 * Get a negated restriction.
+	 *
+	 * @param   Restriction   $other
+	 * @return  Restriction
+	 */
+	public function not(Restriction $other);
+
+	// SOME RESTRICTOINS
+
+	/**
+	 * Get a restriction for an integer.
+	 *
+	 * @return  Restriction
+	 */
+	public function isInt();
+
+	/**
+	 * Get the restriction that some value is larger than $min.
+	 *
+	 * @param   int   $min
+	 * @return  Restriction
+	 */
+	public function greaterThan($min);
+
+	/**
+	 * Get the restriction that some value is smaller then $max.
+	 *
+	 * @param	int   $max
+	 * @return  Restriction
+	 */
+	public function lessThan($max);
+
+	/**
+	 * Get a custom restriction.
+	 *
+	 * If the provided value !$is_ok will either use the $error (if it is a string)
+	 * or provide the value to the $error callback.
+	 *
+	 * @param   callable          $is_ok MUST return boolean
+	 * @param   string|callable   $error
+	 * @return  Restriction
+	 */
+	public function custom($is_ok, $error);
```

Comment:
I do not have a strong opinion versus allowing string and callable or not. Both have benefits. I slightly prefer accepting both for pure convenience on the consumer side.

I think it is necessary to allow custom validation for very specific scenarios. It is a thought one though. The need for a PR for each new validation could lead to not properly validated code though to the fear of the need to discuss a PR in the JF. However, we should always encourage to add reusable validations here. Other than UI stuff, the use of validations are not immediately visible and therefore easier to just leave out if the overhead gets too big.